k2/csrc/cuda/ops.h

```diff
@@ -1,71 +1,142 @@
+  int32_t x = threadIdx.x + blockIdx.x * TILE_DIM;
+  int32_t y = threadIdx.y + blockIdx.y * TILE_DIM;
+
+  for (int32_t i = 0; i < TILE_DIM; i += BLOCK_ROWS) {
+    if (x < cols && (y + i) < rows) {
+      cache[threadIdx.y + i][threadIdx.x] = input[(y + i) * cols + x];
+    }
+  }
+
+  __syncthreads();
+
+  // output index, in a coalesced manner
+  x = threadIdx.x + blockIdx.y * TILE_DIM;
+  y = threadIdx.y + blockIdx.x * TILE_DIM;
+  for (int32_t i = 0; i < TILE_DIM; i += BLOCK_ROWS) {
+    if (x < rows && (y + i) < cols) {
+      output[(y + i) * rows + x] = cache[threadIdx.x][threadIdx.y + i];
+    }
+  }
+}
 
 /*
   Transpose a matrix.  Require src.Size0() == dest.Size1() and src.Size1() ==
   dest.Size0().  This is not the only way to transpose a matrix, you can also
-  do: dest = Array2<T>(src.ToTensor().Transpose(0,1)), which will likely call this
-  function
+  do: dest = Array2<T>(src.ToTensor().Transpose(0,1)), which will likely call
+  this function
 
-     @param [in] c   Context to use, must satisfy `c.IsCompatible(src.Context())` and
-                    `c.IsCompatible(dest->Context())`.
+     @param [in] c   Context to use, must satisfy
+  `c.IsCompatible(src.Context())` and `c.IsCompatible(dest->Context())`.
      @param [in] src  Source array to transpose
-     @param [out] dest  Destination array; must satisfy `dest->Size1() == src.Size0()`
-                     and `dest->Size0() == src.Size1()`.  At exit, we'll have
+     @param [out] dest  Destination array; must satisfy `dest->Size1() ==
+  src.Size0()` and `dest->Size0() == src.Size1()`.  At exit, we'll have
                      dest[i,j] == src[j,i].
  */
-template <typename T> void Transpose(ContextPtr &c,
-                                     Array2<T> &src,
-                                     Array2<T> *dest);
-
+template <typename T>
+void Transpose(ContextPtr &c, const Array2<T> &src, Array2<T> *dest) {
+  assert(c.IsCompatible(src.Context());
+  assert(c.IsCompatible(dest->Context());
+  int32_t rows = src.Dim0();
+  int32_t cols = src.Dim1();
+  // TODO(haowen): limit the number of elements?
+  assert(rows == dest->Dim1());
+  assert(cols == dest->Dim0());
+  const T *src_data = src.Data();
+  T *dest_data = dest->Data();
+  // TODO(haowen): replace if-else with template or inheritance?
+  DeviceType d = c->GetDeviceType();
+  using SumType = typename std::decay<decltype(dest[0])>::type;
+  if (d == kCpu) {
+    for (int i = 0; i < cols; ++i) {
+      for (int j = 0; j < rows; ++j) {
+        output[i * rows + j] = input[j * cols + i];
+      }
+    }
+  } else {
+    assert(d == kCuda);
+    dim3 block_size(TILE_DIM, BLOCK_ROWS, 1);
+    dim3 grid_size((cols + TILE_DIM - 1) / TILE_DIM,
+                   (rows + TILE_DIM - 1) / TILE_DIM);
+    // TODO(haowen): define method `Stream` in context and pass it instead of
+    // the default cuda stream here?
+    TransposeKernel<<<grid_size, block_size>>>(rows, cols, src_data, dest_data);
```

Comment:
In general, after invoking any kernel, we'll want to invoke some kind of macro that checks the return status.
Just FYI.  We don't have the macro yet.  @songmeixu do you have a timeline?

Reply:
... and we need to specify the CUDA stream.  Something like:
```
 TransposeKernel<<<grid_size, block_size, 0, c.GetCudaStream()>>> ...
```

Reply:
ah, there is already such a method in `Context`, will use that, thanks.

Reply:
I add
 https://github.com/danpovey/k2/blob/fc9e9ec26d713fb1f53a2a9cde8326df3c7ba395/k2/csrc/cuda/ops.h#L115
after the kernel call, with `CheckCuda` defined as 
https://github.com/danpovey/k2/blob/fc9e9ec26d713fb1f53a2a9cde8326df3c7ba395/k2/csrc/cuda/ops.h#L22-L30
@songmeixu can update this function later to use the real assert&log system.